File: browser/brave_account/brave_account_service_factory.cc
/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

#include "brave/browser/brave_account/brave_account_service_factory.h"

#include "brave/components/brave_account/brave_account_service.h"
#include "brave/components/brave_account/features.h"
#include "brave/components/sync/service/brave_sync_service_impl.h"
#include "chrome/browser/browser_process.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/sync/sync_service_factory.h"
#include "content/public/browser/browser_context.h"
#include "content/public/browser/storage_partition.h"

namespace brave_account {

// static
BraveAccountServiceFactory* BraveAccountServiceFactory::GetInstance() {
  static base::NoDestructor<BraveAccountServiceFactory> instance;
  return instance.get();
}

// static
BraveAccountService* BraveAccountServiceFactory::GetFor(
    content::BrowserContext* context) {
  CHECK(context);
  return static_cast<BraveAccountService*>(
      GetInstance()->GetServiceForContext(context, true));
}

BraveAccountServiceFactory::BraveAccountServiceFactory()
    : ProfileKeyedServiceFactory("BraveAccountService") {
  CHECK(features::IsBraveAccountEnabled());
}

BraveAccountServiceFactory::~BraveAccountServiceFactory() = default;

bool BraveAccountServiceFactory::ServiceIsCreatedWithBrowserContext() const {
  return true;
}

bool BraveAccountServiceFactory::ServiceIsNULLWhileTesting() const {
  return true;
}

std::unique_ptr<KeyedService>
BraveAccountServiceFactory::BuildServiceInstanceForBrowserContext(
    content::BrowserContext* context) const {
  CHECK(context);
  auto* profile = Profile::FromBrowserContext(context);
  return std::make_unique<BraveAccountService>(
      profile->GetPrefs(),
      context->GetDefaultStoragePartition()
          ->GetURLLoaderFactoryForBrowserProcess(),
      g_browser_process->os_crypt_async(),
      base::BindRepeating(
          [](Profile* profile) -> syncer::BraveSyncServiceImpl* {
            if (!SyncServiceFactory::IsSyncAllowed(profile)) {
              return nullptr;
            }
            return static_cast<syncer::BraveSyncServiceImpl*>(
                SyncServiceFactory::GetForProfile(profile));
          },
          profile));
}

}  // namespace brave_account

File: android/java/org/chromium/chrome/browser/brave_account/BraveAccountServiceFactory.java
/* Copyright (c) 2025 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.brave_account;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.brave_account.mojom.Authentication;
import org.chromium.build.annotations.NullMarked;
import org.chromium.build.annotations.Nullable;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.impl.CoreImpl;

@JNINamespace("chrome::android")
@NullMarked
public class BraveAccountServiceFactory {
    private static final Object sLock = new Object();
    private static @Nullable BraveAccountServiceFactory sInstance;

    public static BraveAccountServiceFactory getInstance() {
        synchronized (sLock) {
            if (sInstance == null) {
                sInstance = new BraveAccountServiceFactory();
            }
        }
        return sInstance;
    }

    private BraveAccountServiceFactory() {}

    public @Nullable Authentication getBraveAccountService(
            Profile profile, @Nullable ConnectionErrorHandler connectionErrorHandler) {
        long nativeHandle =
                BraveAccountServiceFactoryJni.get().getInterfaceToBraveAccountService(profile);
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        if (!handle.isValid()) {
            return null;
        }
        Authentication authentication = Authentication.MANAGER.attachProxy(handle, 0);
        if (connectionErrorHandler != null) {
            Handler handler = ((Interface.Proxy) authentication).getProxyHandler();
            handler.setErrorHandler(connectionErrorHandler);
        }

        return authentication;
    }

    private MessagePipeHandle wrapNativeHandle(long nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        long getInterfaceToBraveAccountService(Profile profile);
    }
}
